### PR TITLE
Add relationship, authentication, reject pending invitation

### PR DESCRIPTION
**Pull request by paulrule**
Postman updated with relationship/add, and get/reject invitation.
Some endpoints now require authentication.
Some cleanup.

### CODE REVIEW

File: backend/typescript/controllers/helpers.ts [hunk 1]
```diff
@@ -89,10 +90,12 @@ export function sendError<T>(res: Response) {
                 res.json(new ErrorResponse(
                     _.values<string>(_.mapValues((error as ValidationError).errors, (v) => v.message))
                 ));
+                logger.error((error as Error).stack);
```
Comment (sundriver) [inline backend/typescript/controllers/helpers.ts]: validationError is not an error
https://github.com/Automattic/mongoose/blob/4.5.0/lib/error/validation.js
correct: (error as ValidationError).stack

Comment (dave-cheong) [inline backend/typescript/controllers/helpers.ts]: Done.

File: commons/RamAPI.ts [hunk 2]
```diff
@@ -225,6 +225,40 @@ export class RelationshipAttribute {
...
+                public unstructuredName:string) {
+    }
+}
+
+export class AttributeDTO {
+    constructor(public code:string,
+                public value:string) {
+    }
+
+    public static build(data:AttributeDTO[]):AttributeDTO[] {
```
Comment (sundriver) [inline commons/RamAPI.ts]: I don't understand, you are getting AttributeDTO[] and returning attributeDTO ? comment please

File: backend/typescript/tests/identity.model.spec.ts [hunk 3]
```diff
@@ -6,65 +6,82 @@ import {
...
-                rawIdValue: 'uuid_1',
-                identityType: IdentityType.LinkId.name,
-                defaultInd: false,
-                linkIdScheme: IdentityLinkIdScheme.MyGov.name,
-                profile: profile1,
-                party: party1
+        Seeder.verbose(false);
+
+        Promise.resolve(null)
+            .then()
```
Comment (sundriver) [inline backend/typescript/tests/identity.model.spec.ts]: then?

File: commons/RamAPI.ts (hunk 2 above)
Comment (dave-cheong) [inline commons/RamAPI.ts]: Fixed casting.

```
    public static build(data:[{code:string, value:string}]):AttributeDTO[] {
        const attributes:AttributeDTO[] = [];
        for (let attribute of data) {
            attributes.push(new AttributeDTO(attribute.code, attribute.value));
        }
        return attributes;
    }
```

File: backend/typescript/server.ts [hunk 4]
```diff
@@ -76,36 +75,30 @@ if (conf.devMode) {
...
+    new RelationshipAttributeNameController(RelationshipAttributeNameModel)
+        .assignRoutes(express.Router()));
+
 server.use('/api/',
-    new RelationshipController(RelationshipModel).assignRoutes(express.Router()));
+    new IdentityController(IdentityModel)
+        .assignRoutes(express.Router()));
 
-// catch 404 and forward to error handler
-server.use((req: express.Request, res: express.Response) => {
```
Comment (sundriver) [inline backend/typescript/server.ts]: why you remove 404 error ?

File: backend/typescript/tests/identity.model.spec.ts (hunk 3 above)
Comment (dave-cheong) [inline backend/typescript/tests/identity.model.spec.ts]: Removed

File: backend/typescript/models/relationship.model.ts [hunk 5]
```diff
@@ -164,9 +165,26 @@ RelationshipSchema.method('toDTO', async function () {
     );
 });
 
+RelationshipSchema.method('rejectPendingInvitation', async function () {
+    if (this.statusEnum() === RelationshipStatus.Pending) {
```
Comment (sundriver) [inline backend/typescript/models/relationship.model.ts]: What if a party has multiple invitation codes, this method rejects all of them.
I thought you reject only the invitation code that is used to establish this relationship.
Do I miss something?

File: backend/typescript/server.ts (hunk 4 above)
Comment (dave-cheong) [inline backend/typescript/server.ts]: It's now back.

File: backend/typescript/models/profile.model.ts [hunk 6]
```diff
@@ -24,16 +24,18 @@ export class ProfileProvider extends RAMEnum {
     public static MyGov = new ProfileProvider('MY_GOV');
     public static SelfAsserted = new ProfileProvider('SELF_ASSERTED');
     public static Vanguard = new ProfileProvider('VANGUARD');
+    public static Temp = new ProfileProvider('TEMP'); // TODO validate what this value should be for temp identities
```
Comment (sundriver) [inline backend/typescript/models/profile.model.ts]: Adding it to tech debt in trello

File: backend/typescript/models/relationship.model.ts (hunk 5 above)
Comment (dave-cheong) [inline backend/typescript/models/relationship.model.ts]: So there is no direct link from Relationship to Identity. Currently, when you do an invite, a temporary party is created which has its own identity. Therefore there is only ever one Identity for a  Pending relationship.

File: backend/typescript/models/profile.model.ts (hunk 6 above)
Comment (dave-cheong) [inline backend/typescript/models/profile.model.ts]: Added to Trello.

File: backend/typescript/models/party.model.ts [hunk 7]
```diff
@@ -78,9 +84,45 @@ PartySchema.method('toDTO', async function () {
...
+        }));
+    }
+
+    // create the relationship
+    const relationship = await RelationshipModel.create({
+        relationshipType: relationshipType,
+        subject: subject.party,
+        subjectNickName: subject.profile.name, // TODO - confirm this
+        delegate: identity.party,
+        delegateNickName: identity.profile.name, // TODO - confirm this
```
Comment (sundriver) [inline backend/typescript/models/party.model.ts]: I assume this is default identity of delegate and subject, could you add this to tech debt

File: backend/typescript/models/party.model.ts [hunk 8]
```diff
@@ -78,9 +84,45 @@ PartySchema.method('toDTO', async function () {
     );
 });
 
+/* tslint:disable:max-func-body-length */
+PartySchema.method('addRelationship', async (dto:RelationshipAddDTO) => {
```
Comment (sundriver) [inline backend/typescript/models/party.model.ts]: is add relationsihp same as accept relationship?

File: backend/typescript/models/party.model.ts (hunk 7 above)
Comment (dave-cheong) [inline backend/typescript/models/party.model.ts]: There is only ever one profile associated with the identity, so as long as the right Identities are supplied from the client, it'll pick the right profile (and hence name). For the delegate there is only one name (the one entered during creation). For the subject though, it depends on how the "business context" is set.

Comment (dave-cheong) [inline backend/typescript/models/party.model.ts]: I'll add a note for discussion in Trello.

File: backend/typescript/models/party.model.ts (hunk 8 above)
Comment (dave-cheong) [inline backend/typescript/models/party.model.ts]: Not the same.

In add relationship - send dto and returns with relationship pointing to temp identity
In review relationship - send invitation code and returns with relationship pointing to temp identity
In accept relationship - send invitation code and returns with relationship pointing to real identity

Comment (dave-cheong) [inline backend/typescript/models/party.model.ts]: Add relationship and Review relationship now available. 
Accept is coming in a later pull request (today).

File: backend/typescript/models/identity.model.ts [hunk 9]
```diff
@@ -19,12 +23,19 @@ const _ProfileModel = ProfileModel;
 /* tslint:disable:no-unused-variable */
 const _PartyModel = PartyModel;
 
-const MAX_PAGE_SIZE = 10;
+const MAX_PAGE_SIZE:number = 10;
```
Comment (sundriver) [inline backend/typescript/models/identity.model.ts]: FYI, typing is inferred so dont need :number

File: backend/typescript/models/party.model.ts (hunk 8 above)
Comment (dave-cheong) [inline backend/typescript/models/party.model.ts]: Added jsdoc

File: backend/typescript/models/identity.model.ts (hunk 9 above)
Comment (dave-cheong) [inline backend/typescript/models/identity.model.ts]: Removed typing.

File: backend/typescript/controllers/relationship.controller.ts [hunk 10]
```diff
@@ -81,10 +99,58 @@ export class RelationshipController {
             .then(sendNotFoundError(res));
     };
 
+    private create = async (req:Request, res:Response) => {
+        const schema = {}; // TODO when DTO is confirmed with front end
```
Comment (sundriver) [inline backend/typescript/controllers/relationship.controller.ts]: technical debt

File: backend/typescript/controllers/relationship.controller.ts [hunk 11]
```diff
@@ -81,10 +99,58 @@ export class RelationshipController {
             .then(sendNotFoundError(res));
     };
 
+    private create = async (req:Request, res:Response) => {
```
Comment (sundriver) [inline backend/typescript/controllers/relationship.controller.ts]: is this create relationship or add relationship?

**Pull request merged**